### PR TITLE
DataChannel subchannels

### LINKED ISSUE #1151: Add subchannels in DataChannels (dataProducer.send() method)
Mechanism to specify which `DataConsumers` should receive a message sent via `dataProducer.send()`.

```ts
const dataConsumer = await transport.consumeData({
  [...],
  subchannels: [ 0, 2, 5 ]
});

const subchannels = dataConsumer.subchannels;
// => [ 0, 2, 5 ] (cloned array)

await dataConsumer.setSubchannels([ 1, 2, 5 ]);
// => Set(4) { 0, 2, 5, 4 }

const subchannels = dataConsumer.subchannels;
// => [ 1, 2, 5 ] (cloned array)

await dataConsumer.setSubchannels([]);
// => [] (cloned array)
```

```ts
// Send a message to all DataConsumers regardless their subchannels.
dataPoducer.send(
  message,
  /* ppid */
  undefined,
  /* subchannels */
  undefined,
  /* requiredSubchannel */
  undefined
);

// Send a message only to DataConsumers in subchannels 1 or 2.
dataPoducer.send(
  message,
  /* ppid */
  undefined,
  /* subchannels */
  [ 1, 2 ],
  /* requiredSubchannel */
  undefined
);

// Send a message only to DataConsumers in subchannels 1 or 2 but just 
// if also in subchannel 0.
dataPoducer.send(
  message,
  /* ppid */
  undefined,
  /* subchannels */
  [ 1, 2 ],
  /* requiredSubchannel */
  0
);
```

### PR DESCRIPTION
**Pull request by ibc**
Fixes #1151

- [x] Node: `transport.consumeData({ subchannels })`.
- [x] Node: `dataConsumer.subchannels` getter.
- [x] Node: `dataConsumer.setSubchannels()` method.
- [x] Node: Proper test in `test-DataConsumer.ts`.
- [x] Node: `dataProducer.send(message, ppid, subchannels, requiredSubchannel)` (new args).
- [x] Node: Proper test for `dataProducer.send()` with `DataConsumers` subscribed to specific subchannels.
- [x] Worker: Make subchannels work (hehe).
- [x] Rust: `transport.consumeData({ subchannels })`.
- [x] Rust: `dataConsumer.subchannels` getter.
- [x] Rust: `dataConsumer.setSubchannels()` method.
- [x] Rust: `dataProducer.send(message, ppid, subchannels, requiredSubchannel)` (new args).
- [x] Rust: Replicate Node tests.

### CODE REVIEW

**Comment by ibc** [comment]
**NOTE:** Fixed, see **UPDATE** at the end of this comment.

Currently dealing with this error:

```
node/src/DataConsumer.ts:693:3 - error TS2322: Type 'number | null' is not assignable to type 'number[]'.
  Type 'null' is not assignable to type 'number[]'.

693   subchannels        : data.subchannels()
      ~~~~~~~~~~~

  node/src/DataConsumer.ts:103:2
    103  subchannels: number[];
         ~~~~~~~~~~~
    The expected type comes from property 'subchannels' which is declared here on type 'DataConsumerDump'

node/src/DataConsumer.ts:693:29 - error TS2554: Expected 1 arguments, but got 0.

693   subchannels        : data.subchannels()
                                ~~~~~~~~~~~~~

  node/src/fbs/data-consumer/dump-response.ts:76:13
    76 subchannels(index: number):number|null {
                   ~~~~~~~~~~~~~
    An argument for 'index' was not provided.

node/src/Transport.ts:1735:59 - error TS2345: Argument of type 'number[]' is not assignable to parameter of type 'number'.

1735   FbsTransport.ConsumeDataRequest.addSubchannels(builder, subchannels);
                                                               ~~~~~~~~~~~
```

I've just added `subchannels` optional array of numbers in `DataConsumerOptions` and mandatory `subchannels` array of numbers in `DumpResponse` in `dataConsumer.fbs`, but this is missing: https://github.com/versatica/mediasoup/pull/1152/files#diff-3253c176609c972f1279e2a6c6ca4c52cbfac8423efcc8b2891b9d8dbd92c358R1733

**UPDATE:** First issue fixed here thanks to @jmillan: https://github.com/versatica/mediasoup/pull/1152/commits/6fca2b37937e76fdd708c938535cec2e5380972b

**Review by jmillan — COMMENTED** [review]
LGTM so far :+1:

**Comment by ibc** [comment]
> LGTM so far 👍

What? Help plz! I don't want kudos I want help

**Comment by ibc** [comment]
**NOTE:** Fixed, see **UPDATE** at the end of this comment.

In this commit https://github.com/versatica/mediasoup/pull/1152/commits/2f5399d3a21700446453ccc3c874e0bf8196fa35 I'm trying to process `subchannels` optional array in `DataConsumer` C++ constructor. However test fails:

```
 FAIL  node/src/tests/test-DataConsumer.ts
  ✕ transport.consumeData() succeeds (2 ms)
  ✕ dataConsumer.dump() succeeds
  ✕ dataConsumer.getStats() succeeds
  ✕ transport.consumeData() on a DirectTransport succeeds (1 ms)
  ✕ dataConsumer.dump() on a DirectTransport succeeds (1 ms)
  ✕ dataConsumer.getStats() on a DirectTransport succeeds
  ✕ dataConsumer.pause() and resume() succeed
  ✕ dataConsumer.close() succeeds (1 ms)
  ✕ Consumer methods reject if closed (1 ms)
  ✕ DataConsumer emits "dataproducerclose" if DataProducer is closed
  ✕ DataConsumer emits "transportclose" if Transport is closed (1 ms)

  ● transport.consumeData() succeeds

    TypeError: FlatBuffers: field 18 must be set

      142 |   builder.requiredField(offset, 6) // data_producer_id
      143 |   builder.requiredField(offset, 8) // type
    > 144 |   builder.requiredField(offset, 18) // subchannels
          |           ^
      145 |   return offset;
      146 | }
      147 |

      at Builder.requiredField (node_modules/flatbuffers/js/builder.js:408:19)
      at Function.endConsumeDataRequest (node/src/fbs/transport/consume-data-request.ts:144:11)
      at createConsumeDataRequest (node/src/Transport.ts:1740:41)
      at PlainTransport.<anonymous> (node/src/Transport.ts:1159:25)
      at node/src/Transport.ts:8:71
      at Object.<anonymous>.__awaiter (node/src/Transport.ts:4:12)
      at PlainTransport.consumeData (node/src/Transport.ts:571:16)
      at node/src/tests/test-DataConsumer.ts:52:35
      at node/src/tests/test-DataConsumer.ts:8:71
      at Object.<anonymous>.__awaiter (node/src/tests/test-DataConsumer.ts:4:12)
      at Object.<anonymous> (node/src/tests/test-DataConsumer.ts:46:53)
```

And if in the first test in `test-DataConsumer.ts` I pass an empty `subchannels: []` in `transport.consumeData()` then I get this different error:

```
  ● transport.consumeData() succeeds

    TypeError: FlatBuffers: object serialization must not be nested.

      126 | static createSubchannelsVector(builder:flatbuffers.Builder, data:number[]|Uint8Array):flatbuffers.Offset;
      127 | static createSubchannelsVector(builder:flatbuffers.Builder, data:number[]|Uint16Array|Uint8Array):flatbuffers.Offset {
    > 128 |   builder.startVector(2, data.length, 2);
          |           ^
      129 |   for (let i = data.length - 1; i >= 0; i--) {
      130 |     builder.addInt16(data[i]!);
      131 |   }

      at Builder.notNested (node_modules/flatbuffers/js/builder.js:244:19)
      at Builder.startVector (node_modules/flatbuffers/js/builder.js:421:14)
      at Function.createSubchannelsVector (node/src/fbs/transport/consume-data-request.ts:128:11)
      at createConsumeDataRequest (node/src/Transport.ts:1733:61)
      at PlainTransport.<anonymous> (node/src/Transport.ts:1159:25)
      at node/src/Transport.ts:8:71
      at Object.<anonymous>.__awaiter (node/src/Transport.ts:4:12)
      at PlainTransport.consumeData (node/src/Transport.ts:571:16)
      at node/src/tests/test-DataConsumer.ts:52:35
      at node/src/tests/test-DataConsumer.ts:8:71
      at Object.<anonymous>.__awaiter (node/src/tests/test-DataConsumer.ts:4:12)
      at Object.<anonymous> (node/src/tests/test-DataConsumer.ts:46:53)
```


**UPDATE:** Fixed in https://github.com/versatica/mediasoup/pull/1152/commits/6b8302725c77b01efdfef2f7be37509fc16047ce

1. Vectors in FBS are required despite `(required)` is not indicated.
2. Vectors must be serialized **before** building the whole FBS request.

File: worker/src/RTC/DataConsumer.cpp [hunk 1]
```diff
@@ -93,24 +93,34 @@ namespace RTC
 	{
 		MS_TRACE();
 
-		flatbuffers::Offset<FBS::SctpParameters::SctpStreamParameters> sctpStreamParametersOffset;
```
Comment (ibc) [inline worker/src/RTC/DataConsumer.cpp:88]: I've removed `Offset` suffix to be consistent. For instance, in `Producer::FillBuffer()` we don't add `Offset` suffix anywhere.

**Comment by ibc** [comment]
@jmillan I'd appreciate if you could review this exact commit: https://github.com/versatica/mediasoup/pull/1152/commits/d4a62fc8dbdbb0fe7ec1a013eb6aba7bf8ebfed9

**Comment by ibc** [comment]
Node and C++ completely done and tests (see [test](https://github.com/versatica/mediasoup/blob/datachannel-subchannels/node/src/tests/test-DirectTransport.ts?ts=2#L249) in `test-DirectTransport.ts`).

So this is ready for review, @jmillan.

File: worker/src/RTC/DataConsumer.cpp (hunk 1 above)
Comment (ibc) [reply worker/src/RTC/DataConsumer.cpp:88]: @jmillan I think we should decide whether `Offset` suffix must be added or not. I see equivalent places where it's added and others in which it's not. IMHO it's redundant and unsightly and we should remove it from everywhere.

**Comment by ibc** [comment]
@nazar-pc maybe you know about this unexpected `cargo clippy` new error?

https://github.com/versatica/mediasoup/actions/runs/6121294807/job/16619834171?pr=1152

```
error[E0793]: reference to packed field is unaligned
    --> C:\Users\runneradmin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ntapi-0.3.7\src\ntexapi.rs:2783:52
     |
2783 |         *tick_count.QuadPart_mut() = read_volatile(&(*USER_SHARED_DATA).u.TickCountQuad);
     |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: packed structs are only aligned by one byte, and many modern architectures penalize unaligned field accesses
     = note: creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)
     = help: copy the field contents to a local variable, or replace the reference with a raw pointer and use `read_unaligned`/`write_unaligned` (loads and stores via `*p` must be properly aligned even when using raw pointers)

error[E0793]: reference to packed field is unaligned
    --> C:\Users\runneradmin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ntapi-0.3.7\src\ntexapi.rs:2807:25
     |
2807 |         ((read_volatile(&(*USER_SHARED_DATA).u.TickCountQuad)
     |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: packed structs are only aligned by one byte, and many modern architectures penalize unaligned field accesses
     = note: creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)
     = help: copy the field contents to a local variable, or replace the reference with a raw pointer and use `read_unaligned`/`write_unaligned` (loads and stores via `*p` must be properly aligned even when using raw pointers)

For more information about this error, try `rustc --explain E0793`.
error: could not compile `ntapi` (lib) due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
Error: The process 'C:\Users\runneradmin\.cargo\bin\cargo.exe' failed with exit code 101
```

**Comment by nazar-pc** [comment]
Looks like it tries to prvent you from doing something bad, is that a new dependency? I don't see any Rust file changes in this PR :confused:

**Comment by ibc** [comment]
> Looks like it tries to prvent you from doing something bad, is that a new dependency? I don't see any Rust file changes in this PR 😕

This PR (which doesn't change absolutely anything in Rust side) has `flatbuffers` branch as target and AFAIS same error happens in `flatbuffers` branch as well: https://github.com/versatica/mediasoup/actions/runs/6111826437/job/16587913166?pr=1064
Not sure if this is due to some recent Rust change/addition of @jmillan in that branch.

**Comment by nazar-pc** [comment]
Hm, didn't check where it comes from, but worth reporting upstream to ntapi crate

File: worker/include/RTC/Router.hpp [hunk 2]
```diff
@@ -96,9 +97,11 @@ namespace RTC
 		void OnTransportDataProducerMessageReceived(
 		  RTC::Transport* transport,
 		  RTC::DataProducer* dataProducer,
-		  uint32_t ppid,
 		  const uint8_t* msg,
-		  size_t len) override;
+		  size_t len,
+		  uint32_t ppid,
+		  std::vector<uint16_t>& subchannels,
```
Comment (jmillan) [inline worker/include/RTC/Router.hpp:103]: This could be `const`, right?

File: worker/include/RTC/Transport.hpp [hunk 3]
```diff
@@ -103,9 +104,11 @@ namespace RTC
 			virtual void OnTransportDataProducerMessageReceived(
 			  RTC::Transport* transport,
 			  RTC::DataProducer* dataProducer,
-			  uint32_t ppid,
 			  const uint8_t* msg,
-			  size_t len) = 0;
+			  size_t len,
+			  uint32_t ppid,
+			  std::vector<uint16_t>& subchannels,
```
Comment (jmillan) [inline worker/include/RTC/Transport.hpp:110]: same, `const`.

File: worker/include/RTC/Transport.hpp [hunk 4]
```diff
@@ -245,17 +248,22 @@ namespace RTC
 			this->DataReceived(len);
 		}
 		void OnDataProducerMessageReceived(
-		  RTC::DataProducer* dataProducer, uint32_t ppid, const uint8_t* msg, size_t len) override;
+		  RTC::DataProducer* dataProducer,
+		  const uint8_t* msg,
+		  size_t len,
+		  uint32_t ppid,
+		  std::vector<uint16_t>& subchannels,
```
Comment (jmillan) [inline worker/include/RTC/Transport.hpp:255]: same.

File: worker/src/RTC/DataConsumer.cpp (hunk 1 above)
Comment (jmillan) [reply worker/src/RTC/DataConsumer.cpp:88]: Yes, I agree.

File: worker/fbs/rtpParameters.fbs [hunk 5]
```diff
@@ -9,11 +9,11 @@ table Boolean {
   value:uint8;
 }
 
-table Integer {
+table Integer32 {
```
Comment (jmillan) [inline worker/fbs/rtpParameters.fbs:12]: even `Int32` could make it.

**Review by jmillan — APPROVED** [review]
Just few considerations noted. :+1:

File: worker/include/RTC/Router.hpp (hunk 2 above)
Comment (ibc) [reply worker/include/RTC/Router.hpp:103]: What you mean? Everything? len, ppid, subchannels and requiredSubchannel?

**Comment by ibc** [comment]
> Hm, didn't check where it comes from, but worth reporting upstream to ntapi crate

Honestly I have no idea what `ntapi` is so no idea about how to report it. I don't even know if this is a bug in `ntapi` or in our code (due to whatever new version of Rust or deps).

File: worker/fbs/rtpParameters.fbs (hunk 5 above)
Comment (ibc) [reply worker/fbs/rtpParameters.fbs:12]: No, please. If so I should also rename `Integer32Array` to `Int32Array`. However `Int32Array` is a JS binary array class so conflicts.

File: worker/include/RTC/Router.hpp (hunk 2 above)
Comment (ibc) [reply worker/include/RTC/Router.hpp:103]: Can we please not open the melon of making all C++ arguments `const`? We don't do it in 99% of cases and honestly I don't feel like if we are missing something useful.
Screenshot 2023-09-11 at 12 05 04

**Comment by ibc** [comment]
> @nazar-pc maybe you know about this unexpected `cargo clippy` new error?
> 
> https://github.com/versatica/mediasoup/actions/runs/6121294807/job/16619834171?pr=1152
> 
> ```
> error[E0793]: reference to packed field is unaligned
>     --> C:\Users\runneradmin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ntapi-0.3.7\src\ntexapi.rs:2783:52
>      |
> 2783 |         *tick_count.QuadPart_mut() = read_volatile(&(*USER_SHARED_DATA).u.TickCountQuad);
>      |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>      |
>      = note: packed structs are only aligned by one byte, and many modern architectures penalize unaligned field accesses
>      = note: creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)
>      = help: copy the field contents to a local variable, or replace the reference with a raw pointer and use 
> ```

So there is a related bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1826485

It seems to be fixed in `ntapi` 0.4.0: https://bugzilla.mozilla.org/show_bug.cgi?id=1826485#c2

However I have no idea which `ntapi` version we are using (whatever `ntapi` is). Output of `cargo tree` doesn't show `ntapi` at all.

Then people say here that _"For some reason (I don't know why) Cargo picked an older version of tokio that used (indirectly) an older version of ntapi. Force Cargo to use new tokio by setting its version to "1.28.0" (the newest version), and it should work."_. We have `tokio` 1.17.0 (whatever `tokio` is) according to `Cargo tree`. But of course we don't specify `tokio` in any `Cargo.toml` so...

**Comment by ibc** [comment]
> > @nazar-pc maybe you know about this unexpected `cargo clippy` new error?
> > https://github.com/versatica/mediasoup/actions/runs/6121294807/job/16619834171?pr=1152
> > ```
> > error[E0793]: reference to packed field is unaligned
> >     --> C:\Users\runneradmin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ntapi-0.3.7\src\ntexapi.rs:2783:52
> >      |
> > 2783 |         *tick_count.QuadPart_mut() = read_volatile(&(*USER_SHARED_DATA).u.TickCountQuad);
> >      |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> >      |
> >      = note: packed structs are only aligned by one byte, and many modern architectures penalize unaligned field accesses
> >      = note: creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)
> >      = help: copy the field contents to a local variable, or replace the reference with a raw pointer and use 
> > ```
> 
> So there is a related bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1826485
> 
> It seems to be fixed in `ntapi` 0.4.0: https://bugzilla.mozilla.org/show_bug.cgi?id=1826485#c2
> 
> However I have no idea which `ntapi` version we are using (whatever `ntapi` is). Output of `cargo tree` doesn't show `ntapi` at all.
> 
> Then people say here that _"For some reason (I don't know why) Cargo picked an older version of tokio that used (indirectly) an older version of ntapi. Force Cargo to use new tokio by setting its version to "1.28.0" (the newest version), and it should work."_. We have `tokio` 1.17.0 (whatever `tokio` is) according to `Cargo tree`. But of course we don't specify `tokio` in any `Cargo.toml` so...

@nazar-pc can we do this in `mediasoup/rust/Cargo.toml`?:
Screenshot 2023-09-11 at 12 27 04

File: worker/include/RTC/Router.hpp (hunk 2 above)
Comment (jmillan) [reply worker/include/RTC/Router.hpp:103]: I just meant `subchannels`.

Comment (ibc) [reply worker/include/RTC/Router.hpp:103]: And why should we make it `const` and not all the other arguments in this method and tons of others?

Comment (jmillan) [reply worker/include/RTC/Router.hpp:103]: But ignore if you want. We can approach the const correctness matter later on.

**Pull request merged**